### ISSUE TITLE
How to dynamically skip over empty layers when performing model speedup after pruning?

### ISSUE THREAD

**Issue by pmmitche** [microsoft/nni#4796]
**Describe the issue**:
When pruning a model at various pruning percentages (10%-95%) using the L1Norm Pruner, I get a `nni.compression.pytorch.speedup.error_code.EmptyLayerError: Pruning a Layer to empty is not legal` error. I was wondering if I can dynamically skip over such layers in these cases? Based on the documentation, I can't determine if a layer will be empty after pruning and before model speedup.

I couldn't find it in the documentation, but I was wondering if there was a way to tell if a layer is empty after pruning and before speedup, so that I can exclude it when speeding up, preventing the EmptyLayerError. Any help would be greatly appreciated, thanks!

**Environment**:
- NNI version: nni==2.7
- Python version: Python 3.8.10
- PyTorch/TensorFlow version: torch==1.10.2+cu113

**How to reproduce it?**:
Prune a model to the point where it gets very small, or start with a small model and continue to prune.

**Comment by J-shang** [comment]
Hello @pmmitche , nni don't have a feature can auto skip empty layer right now, this problem could mainly occur in convolutional layers with `group_num>1`. You could have a try with `pruner = L1NormPruner(..., mode='dependency_aware')`, https://nni.readthedocs.io/en/stable/reference/compression/pruner.html#l1-norm-pruner , in this mode, pruner will try to prune same filter number in each group in one conv layer.

If `mode='dependency_aware'` can not solve your problem, maybe you need to skip these layers in config_list with `{'exclude': True, 'op_names': ['net.conv1']}`, https://nni.readthedocs.io/en/stable/compression/compression_config_list.html#exclude , and we will try to support a way to auto skip these empty layers.

**Comment by pmmitche** [comment]
Awesome! I'll try that out, thank you!

**Comment by pmmitche** [comment]
@J-shang sorry to bother again, but when trying to use the exclude functionality within the config list, the exclude term seems to be ignored. I have tried it on my custom model and on a base timm model of efficientnet and experience the same issue. It seems that the exclude does not prevent NNI from trying to prune the layer, leading to the empty layer exception again. I'm wondering if I'm using it wrong, or if this is an issue with the model or a small bug.

The log output from my minimal example can be found here: https://github.com/pmmitche/Masters-Thesis/blob/main/nni_example_output.txt
(Note, I added some print statements in the source code based on the stack trace to see what layer was causing the issue. Those prints are seen in the log).

A minimal working example that highlights the issue (that the log file is based off of) can be found here: https://github.com/pmmitche/Masters-Thesis/blob/main/minimal_pruning_error_example.py

Thank you for your time and help so far, and any additional help with this would be greatly appreciated!
